### ISSUE TITLE
Expose the subscription id to be compatible with new rpc methods implemented in smoldot

### ISSUE THREAD

**Issue by librelois** [paritytech/jsonrpsee#692]
The new RPC methods [specified here](https://github.com/paritytech/json-rpc-interface-spec/blob/main/src/api/chainHead.md) require to provide as parameter a subscription id that we received previously.

I code a PoC using jsonrpsee to request the smoldot RPC API, and I was forced to fork jsonrpsee to expose the subscription id!

I propose that a new function `get_opaque_subscription_id(&self) -> OpaqueSubscriptionId` be added to the `Subscription` struct.

It is enough that the new type `OpaqueSubscriptionId` implements `Into<JsonValue>`, so it allows to not expose the way to represent this subscription id in jsonrpsee, to be able to modify this implementation detail without breaking the API.

**Comment by niklasad1** [comment]
Hey @librelois 

Just make sure that I understand this correctly, you are saying that this is a client-side issue where the actual `SubscriptionId` is not exposed from subscription API and you need that?!

I see that you need to fetch the subscription ID from the initial call to `chainHead_unstable_follow` to be used for other things however you shouldn't need to access that more than once because you will only get notifications that is equal to your "subscription ID` (that's checked by jsonrpsee)

**Comment by librelois** [comment]
> you are saying that this is a client-side issue where the actual SubscriptionId is not exposed from subscription API and you need that?!

Yes, exactly :) 

> I see that you need to fetch the subscription ID from the initial call to chainHead_unstable_follow to be used for other things however you shouldn't need to access that more than once because you will only get notifications that is equal to your "subscription ID` (that's checked by jsonrpsee)

In fact, we need it several times if we have a local light client, and this is one of the main use cases of smoldot. We call the local light client via its RPC API, and it will open RPC subscriptions with remote full clients.
The light client maintains an internal mapping between the subscriptions ids of the local API and those of the remote APIs, so we have to give it our subscription id each time we call it so that it knows which of its own subscriptions correspond to.

**Comment by librelois** [comment]
> you are saying that this is a client-side issue where the actual SubscriptionId is not exposed from subscription API and you need that?!

Yes, exactly :) 

> however you shouldn't need to access that more than once because you will only get notifications that is equal to your "subscription ID` (that's checked by jsonrpsee)

It seems to me that this is only true in the context of a single RPC subscription. However, to use smoldot RPC api it is necessary to open several subscriptions with different methods, and it seems that smoldot needs to know that these different subscriptions are "linked" (requested by the same application), @tomaka would be better placed to explain why.

**Comment by maciejhirsz** [comment]
@librelois @tomaka 

> The new RPC methods [specified here](https://github.com/paritytech/json-rpc-interface-spec/blob/main/src/api/chainHead.md) require to provide as parameter a subscription id that we received previously.

I find that document incredibly difficult to parse, and I think it might be because there is a glossary conflict there. [JSON-RPC] has:
+ Requests, which includes Notifications (Requests without an `id` that doesn't expect a Response).
+ Responses, which includes Errors.

Putting aside how subscriptions work in jsonrpsee, which is the same way it [has worked in Ethereum](https://geth.ethereum.org/docs/rpc/pubsub), just reading the document I'm left with more questions than answers. Most confusing is:

> 4. Call `chainHead_unstable_call` with `hash` equal to the `finalizedBlockHash` you've just retrieved, `function` equal to `Metadata_metadata`, and an empty `callParameters`.
> 
> 5. If the JSON-RPC server sends back a `{"event": "inaccessible"}` notification, jump back to step 4 to try again. If the JSON-RPC server sends back a `{"event": "error"}` notification, enter panic mode. If the JSON-RPC server sends back a `{"event": "disjoint"}` notification, jump back to step 1. If the JSON-RPC server instead sends back a `{"event": "done"}` notification, save the return value.

What's the response to the call in 4? Are the "notifications" in 5 meant to be responses, or are they actual JSON-RPC Notifications? How does that look in the protocol exactly?

If we need to support free-form Notifications in jsonrpsee (not tied to the pubsub) we can certainly do so.

**Comment by tomaka** [comment]
> just reading the document I'm left with more questions than answers.

The proper link to the documentation is https://paritytech.github.io/json-rpc-interface-spec/api/chainHead.html

What you're reading is an overview. The precise format of everything is described in the individual functions' pages.

**Comment by tomaka** [comment]
> I find that document incredibly difficult to parse, and I think it might be because there is a glossary conflict there. [JSON-RPC](https://www.jsonrpc.org/specification) has:

    Requests, which includes Notifications (Requests without an id that doesn't expect a Response).
    Responses, which includes Errors.

So the idea is that, as the JSON-RPC spec mentions, both sides of a connection can send requests and notifications to each other.

When the client starts a subscription, what it's doing is asking the server to send notifications to it. These notifications are the same and have the same format as the notifications a client could send to a server (I'm saying "could" because in our current APIs we don't have any notification).

**Comment by maciejhirsz** [comment]
@tomaka ye, that's exactly how pubsub has worked since Ethereum, and looking over the specifics of the docs your subscriptions work exactly the same way except for `subscription` being renamed to `subscriptionId` for no good reason?

**Comment by tomaka** [comment]
We can also ask the reverse question: why was `subscription` named `subscription` and not `subscriptionId`? The ethereum doc even uses "subscription id" everywhere in its text. To me `subscriptionId` is just a better word.

There are three solutions to this problem:

- Either we rename `subscriptionId` to `subscription` and admit defeat in the sense that we are forever tied to this arbitrary naming decision some random person has made 5 years ago.
- We modify clients. This is to me the correct option but also obviously the most difficult.
- Decide that `subscription` is actually objectively a better name for this, and rename `subscriptionId` to `subscription`.

Option 2 is to me more sensible in the sense that the new API has many more subscription-like calls, such as [chainHead_unstable_call](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_call.html#chainhead_unstable_call). The reason why these calls are subscription like is to make it possible for the client to receive notifications about the progress of the call, and make it possible for the client to interrupt the call if it's taking too long. It would make sense to expose these calls, despite being subscription-like, with a non-subscription-like API.

In other words, my argument is that this subscription system is too naive anyway, and that further client modifications would be a good idea.

**Comment by maciejhirsz** [comment]
> We can also ask the reverse question: why was `subscription` named `subscription` and not `subscriptionId`? The ethereum doc even uses "subscription id" everywhere in its text. To me `subscriptionId` is just a better word.

The questions aren't even remotely equivalent. It doesn't matter why it was named the way it's named, just at it really doesn't matter whether we drive on the left or the right side of the street, as long as we all drive on the same side of the street. Nobody has ever been confused about what the `"subscription"` field is for, it's perfectly fine. `"subscriptionId"` might be a better name, but changing it is a breaking change in already established ecosystem, it makes the libraries less interoperable and more context specific, there is a real _cost_ to it. It's not solving a "problem", it's just being needlessly pedantic.

> the new API has many more subscription-like calls, such as [chainHead_unstable_call](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_call.html#chainhead_unstable_call) but that might require special processing.

Adding functionality to subscriptions should be totally fine.

**Comment by dvdplm** [comment]
I would be (reluctantly) ok with a solution that lets new clients use `subscriptionId` while old code use `subscription`. There are several downsides to that (if it's even possible in a somewhat ergonomic way) but I'd be prepared to swallow it. So yeah, both option 1 and 2. Possible? Serde alias something something? Or `#[serde(rename = "…")]`?

**Comment by niklasad1** [comment]
After some thought, I agree with Maciej here that renaming `"subscription"` to `"subscriptionId"` is not worth the effort.

The [ethereum pubsub spec](https://geth.ethereum.org/docs/rpc/pubsub) is widely adopted and doesn't make sense to diverge from it for just renaming the `"subscription"` most users will not care about this field at all.

Further, it will be quite nasty support such as configurable `subscription` field in the servers via `configuration flag` or something.
In the client's it would be easier we could just to `deserialize` any field and then forward it the users to handle but I don't think it's worth the effort because of how painful that API would be to use.

[Parity's old jsonrpc library](https://github.com/paritytech/jsonrpc/) is widely used and have "hardcoded" the `"subscription"` field too.

@librelois 

We would accept a change to expose the subscriptionId from the `Subscription struct`, if you have something working feel free to open a PR :)

**Comment by niklasad1** [comment]
I just realized that this is already supported https://docs.rs/jsonrpsee-core/0.13.1/jsonrpsee_core/client/struct.Subscription.html#method.kind

Unfortunately, you have to match on the SubscriptionKind because we also support subscription per method name.
